### PR TITLE
Make all format names case insensitive.

### LINKED ISSUE #59526: Consider making all format names case insensitive.
**Use case**

Let's just give up and accept if a user writes Tsv, or TSV, or tsv, or even rowbinary.

### PR DESCRIPTION
**Pull request by HowePa**
close #59526

### Changelog category:
- New Feature

### Changelog entry:
Make all format names case insensitive, like Tsv, or TSV, or tsv, or even rowbinary.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 52ed022187aa08852894702d324a6310cc437bc8 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59560/52ed022187aa08852894702d324a6310cc437bc8/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

File: src/Formats/registerFormats.cpp [hunk 1]
```diff
@@ -291,6 +291,8 @@ void registerFormats()
     registerDWARFSchemaReader(factory);
     registerOneSchemaReader(factory);
     registerNpySchemaReader(factory);
+
+    factory.registerAllFormatNames();
```
Comment (Algunenano) [inline src/Formats/registerFormats.cpp:295]: This is wrong for 2 reasons:
* It's exactly backwards from the dozens of previous calls
* The formats are already registered. If we need another structure to do lookups, add it and change `registerInputFormat` / `registerRandomAccessInputFormat`/ whatever.

File: tests/queries/0_stateless/00309_formats.sql [hunk 2]
```diff
@@ -7,3 +7,6 @@ SELECT number * 246 + 10 AS n, toDate('2000-01-01') + n AS d, range(n) AS arr, a
 SELECT number * 246 + 10 AS n, toDate('2000-01-01') + n AS d, range(n) AS arr, arrayStringConcat(arrayMap(x -> reinterpretAsString(x), arr)) AS s, (n, d) AS tuple FROM system.numbers LIMIT 2 FORMAT JSON;
 SELECT number * 246 + 10 AS n, toDate('2000-01-01') + n AS d, range(n) AS arr, arrayStringConcat(arrayMap(x -> reinterpretAsString(x), arr)) AS s, (n, d) AS tuple FROM system.numbers LIMIT 2 FORMAT JSONCompact;
 SELECT number * 246 + 10 AS n, toDate('2000-01-01') + n AS d, range(n) AS arr, arrayStringConcat(arrayMap(x -> reinterpretAsString(x), arr)) AS s, (n, d) AS tuple FROM system.numbers LIMIT 2 FORMAT XML;
+SELECT number * 246 + 10 AS n, toDate('2000-01-01') + n AS d, range(n) AS arr, arrayStringConcat(arrayMap(x -> reinterpretAsString(x), arr)) AS s, (n, d) AS tuple FROM system.numbers LIMIT 2 FORMAT rowbinary;
```
Comment (Algunenano) [inline tests/queries/0_stateless/00309_formats.sql:10]: Please use a different file to add new tests. That way it's easier to identify changes.

Also, we need to add tests (and probably code) for other functions that support passing formats. For example:
```
SELECT * FROM file('/tmp/a', 'Csv')
```

Look into callers to FormatFactory's getInput/getOutput*, but probably we just need to accept any case in those methods.

File: tests/queries/0_stateless/02882_formatQuery.reference [hunk 3]
```diff
@@ -26,6 +26,9 @@ SELECT 1
 SELECT 1
 1
 1
+SELECT 1 FORMAT TSV
```
Comment (Algunenano) [inline tests/queries/0_stateless/02882_formatQuery.reference:29]: Same, let's move all the tests to a different file or files, please

**Review by Algunenano — CHANGES_REQUESTED** [review]

File: src/Formats/registerFormats.cpp (hunk 1 above)
Comment (HowePa) [reply src/Formats/registerFormats.cpp]: I observed that `FormatFactory::file_extension_formats` can also be used as a case-insensitive format name mapping, so I implemented `FormatFactory::getOriginalFormatNameIfExists` to use it to look up farmat names. In addition, I created `KnownFormatNames::case_insensitive_names_map` to get the format name when parsing the FORMAT.

File: tests/queries/0_stateless/00309_formats.sql (hunk 2 above)
Comment (HowePa) [reply tests/queries/0_stateless/00309_formats.sql]: Added a new test `00309_formats_case_insensitive.sql`.

File: tests/queries/0_stateless/02882_formatQuery.reference (hunk 3 above)
Comment (HowePa) [reply tests/queries/0_stateless/02882_formatQuery.reference]: Added a new test `00309_formats_case_insensitive.sql`.

File: src/Parsers/ExpressionElementParsers.cpp [hunk 4]
```diff
@@ -242,6 +243,39 @@ bool ParserIdentifier::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
 }
 
 
+bool ParserFormatIdentifier::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
```
Comment (Algunenano) [inline src/Parsers/ExpressionElementParsers.cpp:246]: I don't understand why should the parser know about available formats. The factory needs to know how to handle it, not the parser.

File: src/Formats/FormatFactory.cpp [hunk 5]
```diff
@@ -777,13 +786,13 @@ String FormatFactory::getAdditionalInfoForSchemaCache(const String & name, Conte
 
 bool FormatFactory::isInputFormat(const String & name) const
 {
-    auto it = dict.find(name);
+    auto it = dict.find(getOriginalFormatNameIfExists(name));
```
Comment (Algunenano) [inline src/Formats/FormatFactory.cpp:789]: If all formats are case insensitive, we should just store their lowercase value in the hash table and look for lowercase here. No need of `getOriginalFormatNameIfExists`

File: src/Parsers/ExpressionElementParsers.cpp (hunk 4 above)
Comment (HowePa) [reply src/Parsers/ExpressionElementParsers.cpp]: format
I want this effect, although it doesn't affect usability. I have removed it.

File: src/Formats/FormatFactory.cpp (hunk 5 above)
Comment (HowePa) [reply src/Formats/FormatFactory.cpp]: Now, it's just converted to lowercase.

Comment (HowePa) [reply src/Formats/FormatFactory.cpp:789]: 1. I found that using lowercase directly causes the name column in the `system.forms` table to be all lowercase, in `StorageSystemFormats::fillData()`
2. Many of the functional tests use name for queries (such as, `select number_of_rows from system.schema_inference_cache where format='Npy'` in 02908_Npy_files_caching), so it is necessary to use original format name

**Review by Algunenano — APPROVED** [review]
Can you merge with current master please?

**Comment by HowePa** [comment]
> Can you merge with current master please?

replace to #60420

**Pull request closed without merging**